java/arcs/android/storage/service/SuspendingServiceMethods.kt

```diff
@@ -0,0 +1,82 @@
+package arcs.android.storage.service
+
+import android.os.IInterface
+import arcs.core.crdt.CrdtException
+import arcs.core.util.Log
+import kotlin.coroutines.resumeWithException
+import kotlinx.coroutines.CancellableContinuation
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.suspendCancellableCoroutine
+
+/**
+ * Suspends the current coroutine. The caller is provided with an [IRegistrationCallback] which
+ * can be used to signal resumption; most likely you are passing this to the corresponding
+ * Android storage service method call.
+ */
+@ExperimentalCoroutinesApi
+suspend fun suspendForRegistrationCallback(block: (IRegistrationCallback) -> Unit) =
+    suspendCancellableCoroutine<Int> { block(ContinuationRegistrationCallback(it)) }
+
+/**
+ * Suspends the current coroutine. The caller is provided with an [IResultCallback] which
+ * can be used to signal resumption; most likely you are passing this to the corresponding
+ * Android storage service method call.
+ */
+@ExperimentalCoroutinesApi
+suspend fun suspendForResultCallback(block: (IResultCallback) -> Unit) =
+    suspendCancellableCoroutine<Boolean> { block(ContinuationResultCallback(it)) }
+
+/**
+ * Helper that converts a continuation expecting an [Int] into an [IRegistrationCallback] that will
+ * complete the continuation when [onSuccess] is called, or raise an exception when [onFailure] is
+ * called.
+ */
+@ExperimentalCoroutinesApi
+class ContinuationRegistrationCallback(
+    private val continuation: CancellableContinuation<Int>
+) : IRegistrationCallback by object : IRegistrationCallback.Stub() {
```

Comment:
Curious why you chose to do implementation delegation, rather than just extending the Stub?

Reply:
Ah, good catch. Artifact of earlier iteration in which I had a common base class that I was extending. That's gone, so I can clean this up.